charmcraft/models/config.py

```diff
@@ -0,0 +1,286 @@
+# Copyright 2023 Canonical Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+# http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# For further info, check https://github.com/canonical/charmcraft
+
+"""Charmcraft configuration pydantic model."""
+
+import datetime
+import pathlib
+from typing import Optional, List, Dict, Any
+
+import pydantic
+
+from craft_cli import CraftError
+from charmcraft.parts import process_part_config
+from charmcraft.utils import get_host_architecture
+from charmcraft.format import format_pydantic_errors
+
+
+class ModelConfigDefaults(
+    pydantic.BaseModel, extra=pydantic.Extra.forbid, frozen=True, validate_all=True
+):
+    """Define Charmcraft's defaults for the BaseModel configuration."""
+
+
+class CustomStrictStr(pydantic.StrictStr):
+    """Generic class to create custom strict strings validated by pydantic."""
+
+    @classmethod
+    def __get_validators__(cls):
+        """Yield the relevant validators."""
+        yield from super().__get_validators__()
+        yield cls.custom_validate
```

Comment:
Did this get moved from elsewhere or added?

If added: This technically makes the class abstract, right? (Since there's no `custom_validate` implementation?) If so, please add an abstract method for `custom_validate` that we override below.

(If it's just been moved, I'll make an issue for it so we can do that cleanup separately)

Reply:
This one is moved, I did not touch the code.

Reply:
Thanks! I've made an issue for it. https://github.com/canonical/charmcraft/issues/1118